### PR TITLE
Add a new post-processor to restrict recommendations based on replica count

### PR DESCRIPTION
**Pull request by iksaif**
See https://datadoghq.atlassian.net/wiki/spaces/AG/pages/2971205765/WPA+VPA+Cooperation

### CODE REVIEW

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+		if value == "" {
+			watermarks = []int{10, 100}
+		} else if err := json.Unmarshal([]byte(value), &watermarks); err != nil {
+			klog.Errorf("Skipping restrictions definition '%s' in vpa %s/%s due to bad format, error:%#v", value, vpa.ID.Namespace, vpa.ID.VpaName, err)
+		}
+		if len(watermarks) != 2 {
+			klog.Errorf("Skipping restrictions definition '%s' in vpa %s/%s due to bad format", value, vpa.ID.Namespace, vpa.ID.VpaName)
+			continue
+		}
```
Comment (zaymat) [inline vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:98]: In both these cases, we should probably reset `watermarks` to `int{}` and return early. For example, if a user gives an input with one or three numbers, i think we'll trigger the error, but we'll still return a none empty array.

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+// Any increase of resources is considered an upscale, any decrease of resources is considered a downscale.
+// This is best used with the `maintainedRatio` policy or with a single controlled resource.
+type ReplicaRestrictionsPostProcessor struct {
+	ControllerFetcher controllerfetcher.ControllerFetcher
+}
+
+// Process applies the Resource Ratio post-processing to the recommendation.
+func (r *ReplicaRestrictionsPostProcessor) Process(vpa *model.Vpa, recommendation *vpa_types.RecommendedPodResources, _ *vpa_types.PodResourcePolicy) *vpa_types.RecommendedPodResources {
+	watermarks := readWatermarksFromVPAAnnotations(vpa)
+	if watermarks == nil || len(watermarks) == 0 || recommendation == nil {
```
Comment (zaymat) [inline vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:55]: probably worth checking that the array is of length 2

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+// Apply returns a recommendation for the given pod, adjusted to obey maintainedRatio policy
+func (r *ReplicaRestrictionsPostProcessor) apply(
+	podRecommendation *vpa_types.RecommendedPodResources,
+	replicas, lowWatermark, highWatermark int,
+	pod *apiv1.Pod) (*vpa_types.RecommendedPodResources, error) {
+
+	if podRecommendation == nil {
+		// If there is no recommendation let's skip that post-processor.
+		return nil, nil
+	}
```
Comment (zaymat) [inline vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:116]: I think we can skip even earlier in the post processor

File: vertical-pod-autoscaler/examples/hamster-external-metrics.yaml [hunk 4]
```diff
@@ -12,6 +12,7 @@ metadata:
   annotations:
     vpa.datadoghq.com/metric-cpu-hamster: datadogmetric@default:hamster-cpu
     vpa.datadoghq.com/metric-memory-hamster: datadogmetric@default:hamster-memory
+    vpa-post-processor.kubernetes.io/replicaRestrictions: "[5, 10]"
```
Comment (dbenque) [inline vertical-pod-autoscaler/examples/hamster-external-metrics.yaml:15]: nit (UX): would call it replicaRange

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+// isUpscale: detects if the recommendation is an upscale. Any increase of resources is considered as upscale.
+func isUpscale(recommendation apiv1.ResourceList, containerOriginalResources apiv1.ResourceList) bool {
+	for recommendationType, _ := range recommendation {
+		q := containerOriginalResources.Name(recommendationType, resource.DecimalSI).MilliValue()
+		r := recommendation.Name(recommendationType, resource.DecimalSI).MilliValue()
+		if r > q {
+			return true
+		}
```
Comment (zaymat) [inline vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:182]: Nit: you can use `return r.Cmp(q) == 1` instead.

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+// isDownscale: detects if the recommendation is a downscale. Any decrease of resources is considered as downscale.
+func isDownscale(recommendation apiv1.ResourceList, containerOriginalResources apiv1.ResourceList) bool {
+	for recommendationType := range recommendation {
+		q := containerOriginalResources.Name(recommendationType, resource.DecimalSI).MilliValue()
+		r := recommendation.Name(recommendationType, resource.DecimalSI).MilliValue()
+		if r < q {
+			return true
+		}
```
Comment (zaymat) [inline vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:194]: Nit: same as above

**Review by zaymat — COMMENTED** [review]
Left a few comments on the code. 

More generally, what's the plan for this first iteration? Is the goal to set these values manually at first, or is it planed to have a controller that set this annotation on the pods?

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+	}
+	return false
+}
+
+// applyRestrictedScalingVPAPolicy makes sure we do not make a recommendation if we are restricted.
+func applyRestrictedScalingVPAPolicy(recommendation apiv1.ResourceList, isRestricted bool, containerOriginalResources apiv1.ResourceList) {
+
+	for recommendationType := range recommendation {
+		if isRestricted {
```
Comment (dbenque) [inline vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:203]: While I think code is correct, I cannot understand how this solution is going to be stable.
Based on an instant value that is the replica count (that is influenced by horizontal scaling) you can suddenly go back to chart initial value! While the jump might me contained with the `triggered` mode due to the chart mutation, this can work well with classic mode `auto` and `recreate`.
I can see the recommendation slowly moving from an initial value toward a target, potentially with big difference, and suddenly going back to what the chart was defining.

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+		if key != vpaPostProcessorPrefix + vpaPostProcessorReplicaRestrictionsSuffix {
+			continue
+		}
+
+		if value == "" {
+			watermarks = []int{10, 100}
+		} else if err := json.Unmarshal([]byte(value), &watermarks); err != nil {
+			klog.Errorf("Skipping restrictions definition '%s' in vpa %s/%s due to bad format, error:%#v", value, vpa.ID.Namespace, vpa.ID.VpaName, err)
+		}
+		if len(watermarks) != 2 {
```
Comment (dbenque) [inline vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:95]: also check than first value is lower that second

**Comment by iksaif** [comment]
> More generally, what's the plan for this first iteration? Is the goal to set these values manually at first, or is it planed to have a controller that set this annotation on the pods?

Here's the plan https://datadoghq.atlassian.net/wiki/spaces/AG/pages/2971205765/WPA+VPA+Cooperation#Iteration-0

TL;DR; If it works by hand, we can see about fetching the values from the associated WPA (but I think it's not super important since it's quite easy to do in a chart)

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go (hunk 7 above)
Comment (iksaif) [reply vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:203]: https://datadoghq.atlassian.net/wiki/spaces/AG/pages/2971205765/WPA+VPA+Cooperation#Iteration-0

Yes we should not look at instant metrics in the final solution, but it's probably still good enough to start since services usually don't jump from 0 to 100 pods and back to 0 every 10 seconds

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+// ReplicaRestrictionsPostProcessor ensures that defined ratio constraints between resources is applied.
+// The definition is done via annotation on the VPA object with format: vpa-post-processor.kubernetes.io/replicaRestrictions="[lowWatermark, highWatermark]"
+// - lowWatermark is the maximum number of replicas to allow a vertical downscale
+// - highWatermark is the minimum number of replicas to allow a vertical upscale
+// In other words:
+// - If we have less than <lowWatermark> replicas, we will allow the pods to be smaller, the intent it to keep the number
+//   of pods above <lowWatermark>. This allows you to improve resource utilisation to avoid having few very big and underutilised pods.
+// - If we have more than <highWatermark> replicas, we will allow the pods to be bigger, the intent it to keep the number
+//   of pods below <highWatermark>. This allows you to reduce the pressure on the k8s control plane.
+// If the annotation is defined and set to an empty value, the default value is used: [10, 100].
```
[the remark this answers is not in the document]
Comment (iksaif) [reply vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:45]: Which is how it works, it justs allow the annotation to be present but empty.

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+		ApiVersion: vpa.TargetRef.APIVersion,
+	})
+	if err != nil {
+		klog.Errorf("Failed to apply ReplicaRestrictionsPostProcessor (controller fetch) to vpa %s/%s due to error: %#v", vpa.ID.Namespace, vpa.ID.VpaName, err)
+		return recommendation
+	}
+
+	pod := newPodFromTemplate(podTemplate)
+
+	updatedRecommendation, err := r.apply(recommendation, vpa.PodCount, lowWatermark, highWatermark, pod)
```
[the remark this answers is not in the document]
Comment (iksaif) [reply vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:75]: I agree this is the plan for the second iteration https://datadoghq.atlassian.net/wiki/spaces/AG/pages/2971205765/WPA+VPA+Cooperation#Iteration-1

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+	if isRestricted {
+		klog.V(2).Infof("[%s] Scaling is restricted %d - replica restrictions [%d, %d] - upscale: %t - downscale: %t", container.Name, replicas, lowWatermark, highWatermark, isAnUpscale, isADownscale)
+	} else {
+		klog.V(4).Infof("[%s] Scaling is not restricted %d - replica restrictions [%d, %d] - upscale: %t - downscale: %t", container.Name, replicas, lowWatermark, highWatermark, isAnUpscale, isADownscale)
+	}
+	return isRestricted
+}
+
+// isUpscale: detects if the recommendation is an upscale. Any increase of resources is considered as upscale.
+func isUpscale(recommendation apiv1.ResourceList, containerOriginalResources apiv1.ResourceList) bool {
```
[the remark this answers is not in the document]
Comment (iksaif) [reply vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:176]: This is not making any assumption:
- anything that increases any kind of resource is an upscale
- anything that decreases any kind of resource is a downscale

This means that downscales and upscales are not mutually exclusive

Obviously this has some complex implications when you control multiple resources, and it probably works well only if you use the post-processor to maintain the ratio. But unless you use a fixed ratio I don't see how having VPA+WPA on the same resource would work

[the remark this answers is not in the document]
Comment (dbenque) [reply vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:176]: sorry, I misunderstood on first read, the return false is only in case you did not find any upscale.
That is fine

File: vertical-pod-autoscaler/examples/hamster-external-metrics.yaml (hunk 4 above)
Comment (iksaif) [reply vertical-pod-autoscaler/examples/hamster-external-metrics.yaml:15]: fixed

File: vertical-pod-autoscaler/pkg/recommender-external/main.go [hunk 12]
```diff
@@ -53,9 +53,11 @@ var (
 // Post processors flags
 var (
 	// CPU as integer to benefit for CPU management Static Policy ( https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/#static-policy )
-	postProcessorCPUasInteger = flag.Bool("cpu-integer-post-processor-enabled", false, "Enable the cpu-integer recommendation post processor. The post processor will round up CPU recommendations to a whole CPU for pods which were opted in by setting an appropriate label on VPA object (experimental)")
+	postProcessorCPUasInteger = flag.Bool("cpu-integer-post-processor-enabled", true, "Enable the cpu-integer recommendation post processor. The post processor will round up CPU recommendations to a whole CPU for pods which were opted in by setting an appropriate label on VPA object (experimental)")
```
[the remark this answers is not in the document]
Comment (iksaif) [reply vertical-pod-autoscaler/pkg/recommender-external/main.go:56]: ok

File: vertical-pod-autoscaler/pkg/recommender-external/main.go [hunk 13]
```diff
@@ -53,9 +53,11 @@ var (
...
-	postProcessorMaintainResourceRatio = flag.Bool("resource-ratio-post-processor-enabled", false, "Enable the resource-ratio recommendation post processor. The post processor will ensure that resource ratio is maintain for pods which were opted in by setting an appropriate label on VPA object (experimental)")
+	postProcessorMaintainResourceRatio = flag.Bool("resource-ratio-post-processor-enabled", true, "Enable the resource-ratio recommendation post processor. The post processor will ensure that resource ratio is maintain for pods which were opted in by setting an appropriate label on VPA object (experimental)")
+	// Replica restrictions post-processor.
+	postProcessorReplicaRestrictions = flag.Bool("replica-restrictions-post-processor-enabled", true, "Enable the replica-restrictions recommendation post processor. The post processor will ensure that the allow upscales/downscales based on the currently running number of replicas for controllers which were opted in by setting an appropriate label on VPA object (experimental)")
```
[the remark this answers is not in the document]
Comment (iksaif) [reply vertical-pod-autoscaler/pkg/recommender-external/main.go:60]: ok

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+	controllerfetcher "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/model"
+)
+
+const (
+	vpaPostProcessorReplicaRestrictionsSuffix = "replicaRestrictions"
+)
+
+// ReplicaRestrictionsPostProcessor ensures that defined ratio constraints between resources is applied.
+// The definition is done via annotation on the VPA object with format: vpa-post-processor.kubernetes.io/replicaRestrictions="[lowWatermark, highWatermark]"
```
[the remark this answers is not in the document]
Comment (iksaif) [reply vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:37]: changing this

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go (hunk 2 above)
Comment (iksaif) [reply vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:55]: done

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go (hunk 8 above)
Comment (iksaif) [reply vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:95]: done

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+		}
+
+		if value == "" {
+			watermarks = []int{10, 100}
+		} else if err := json.Unmarshal([]byte(value), &watermarks); err != nil {
+			klog.Errorf("Skipping restrictions definition '%s' in vpa %s/%s due to bad format, error:%#v", value, vpa.ID.Namespace, vpa.ID.VpaName, err)
+		}
+		if len(watermarks) != 2 {
+			klog.Errorf("Skipping restrictions definition '%s' in vpa %s/%s due to bad format", value, vpa.ID.Namespace, vpa.ID.VpaName)
+			continue
```
[the remark this answers is not in the document]
Comment (iksaif) [reply vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:97]: done

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go (hunk 1 above)
Comment (iksaif) [reply vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:98]: changed

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go (hunk 3 above)
Comment (iksaif) [reply vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:116]: done

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go (hunk 5 above)
Comment (iksaif) [reply vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:182]: I never understand how `Cmp` works, I find a basic comparison easier to read

File: vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go (hunk 7 above)
Comment (dbenque) [reply vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go:216]: fixed/mitigated thanks to check in the process function `*vpa.UpdateMode != vpa_types.UpdateModeTrigger`

**Review by dbenque — APPROVED** [review]
